src/broker/broker.c

```diff
@@ -1965,13 +1967,75 @@ static void module_status_cb (module_t *p, int prev_status, void *arg)
     }
 }
 
+static bool signal_is_deadly (int signum)
+{
+    int deadly_sigs[] = { SIGHUP, SIGINT, SIGQUIT, SIGTERM, SIGALRM };
+    for (int i = 0; i < ARRAY_SIZE (deadly_sigs); i++) {
+        if (signum == deadly_sigs[i])
+            return true;
+    }
+    return false;
+}
+
+static void killall_cb (flux_future_t *f, void *arg)
+{
+    broker_ctx_t *ctx = arg;
+    int count = 0;
+    if (flux_rpc_get_unpack (f, "{s:i}", "count", &count) < 0) {
+        flux_log_error (ctx->h,
+                        "job-manager.killall: %s",
+                        future_strerror (f, errno));
+    }
+    flux_future_destroy (f);
+    if (count) {
+        flux_log (ctx->h,
+                  LOG_INFO,
+                  "forwarded signal %d to %d jobs",
+                  (int) ptr2int (flux_future_aux_get (f, "signal")),
+                  count);
+    }
+}
+
+static int killall_jobs (broker_ctx_t *ctx, int signum)
+{
+    flux_future_t *f = NULL;
+    if (!(f = flux_rpc_pack (ctx->h,
+                    "job-manager.killall",
+                    FLUX_NODEID_ANY,
+                    0,
+                    "{s:b s:i s:i}",
+                    "dry_run", 0,
+                    "userid", FLUX_USERID_UNKNOWN,
+                    "signum", signum))
+        || flux_future_then (f, -1., killall_cb, ctx) < 0) {
+            flux_future_destroy (f);
+            return -1;
+    }
+    if (flux_future_aux_set (f, "signum", int2ptr (signum), NULL) < 0)
+        flux_log_error (ctx->h, "killall: future_aux_set");
+    return 0;
+}
+
 static void signal_cb (flux_reactor_t *r, flux_watcher_t *w,
-                         int revents, void *arg)
+                       int revents, void *arg)
 {
     broker_ctx_t *ctx = arg;
     int signum = flux_signal_watcher_get_signum (w);
 
     flux_log (ctx->h, LOG_INFO, "signal %d", signum);
+
+    if (ctx->rank == 0 && !signal_is_deadly (signum)) {
+        /* Attempt to forward non-deadly signals to jobs. If that fails,
+         * the fall through to state_machine_kill() so the signal is
+         * delivered somewhere.
+         */
+        if (killall_jobs (ctx, signum) == 0)
+            return;
```

Comment:
Might be good to note that this is expected to fail immediately on rank 0 because of the way the broker's internal flux_t handle is implemented.

Aside: is it intended to forward signals to rank 0 if one is received on a rank > 0 broker?

Reply:
It should only fail if there is no job-manager module loaded, which should be the extremely uncommon case.
Oh, do you mean instead of getting ENOSYS from the RPC response?

Note: signals are only forwarded from rank 0.

Reply:
Right -  I just meant it might be good to note that the failure, if it occurs, is immediate on rank 0 due to that quirk.  On rank > 0, the request would be forwarded upstream, which led to my wondering whether this code would ever be called on rank > 0.

Reply:
This code is currently within a block `if (ctx->rank == 0 && !signal_is_deadly (signum))`, so it will only be called on rank 0. (Still not understanding your question or request here maybe?)

Reply:
FYI - changing the comments to the following
```diff
+
+    if (ctx->rank == 0 && !signal_is_deadly (signum)) {
+        /* Attempt to forward non-deadly signals to jobs. If that fails,
+         * then fall through to state_machine_kill() so the signal is
+         * delivered somewhere.
+         */
+        if (killall_jobs (ctx, signum) == 0)
+            return;
+        /*
+         * Note: flux_rpc(3) in the rank 0 broker to the job manager module
+         *  is expected to fail immediately if the job-manager module is not
+         *  loaded due to the broker internal flux_t handle implementation.
+         */
+        flux_log (ctx->h,
+                  LOG_INFO,
+                  "killall failed, delivering signal %d locally instead",
+                  signum);
+    }
```

Reply:
That's what I wanted.  Thank you!